### PR TITLE
Setup testing infrastructure

### PR DESCRIPTION
**Pull request by meeh0w**
## Overview
* breaks down existing code into testable modules
* adds Jest and tests to the service
* adds codecov to report on code coverage impact for each pull request

## Checklist

- [x] All git commits are signed. (REQUIRED)
- [ ] All new methods or updated methods have clear docstrings.
- [x] Testing added or updated for new methods.
- [ ] Verified if any changes impact the WebPortal Health Checks.
- [ ] Appropriate documentation updated.
- [ ] Changelog file created.

## Issues Closed
SKY-692

### CODE REVIEW

**Review by MSevey — DISMISSED** [review]
It would be nice as at least a follow up to add some inline comments to the test files to explain the scope of the tests to help reviewers ensure that the tests are covering all the edge cases.

File: package.json [hunk 1]
```diff
@@ -8,11 +8,17 @@
     "is-valid-domain": "^0.1.6"
   },
   "devDependencies": {
+    "@types/jest": "^27.5.1",
```
Comment (interstella-5555) [inline package.json]: it should be safe to remove, we're not running typescript on this repo

File: package.json [hunk 2]
```diff
@@ -8,11 +8,17 @@
...
+    "codecov": "^3.8.3",
+    "jest": "^28.1.0",
+    "node-fetch": "2",
```
Comment (interstella-5555) [inline package.json:14]: why version 2 when latest is 3.2.5 ?

File: src/resolver.test.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    });
+  });
+
+  describe(".resolve", () => {
+    it("invokes dns.resolveTxt with proper params", async () => {
+      dns.resolveTxt.mockImplementationOnce(() => {});
+      resolver.resolve("skynetlabs.com");
+
+      expect(dns.resolveTxt).toHaveBeenCalledWith("_dnslink.skynetlabs.com", expect.any(Function));
+    });
```
Comment (interstella-5555) [inline src/resolver.test.js:64]: I think this is an implementation detail, might not be too useful to test it separately

File: src/resolver.test.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+      });
+    });
+
+    describe("when TXT records contain multiple skynet dnslinks", () => {
+      beforeEach(() => {
+        dns.resolveTxt.mockImplementationOnce(mockedDnsResolveTxt(null, FIXTURES.MULTIPLE_SKYNET_LINKS));
+      });
+
+      it("throws MultipleSkylinksError", () => {
+        expect(resolver.resolve("multiple-skylinks.com")).rejects.toThrow(MultipleSkylinksError);
```
Comment (interstella-5555) [inline src/resolver.test.js]: You could define `const validDomain = "example.com";` on top of the file and use it as param to `resolve` in all tests where valid domain is required to simplify the tests by making them more similar (different only by the things that really matter).

File: src/server.test.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  beforeAll(() => {
+    app.start(1234);
+  });
+
+  afterAll(() => {
+    app.server.close();
+  });
+
+  describe("/dnslink/:domainName route", () => {
+    describe("when domain is configured with a skylink", () => {
```
Comment (interstella-5555) [inline src/server.test.js:17]: we're missing a test for this endpoint returning skylink with sponsor key

File: src/resolver.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            `Multiple TXT records with valid sponsor key found for ${lookup}, only one allowed`
+          );
+        }
+
+        if (sponsors.length === 1) {
+          // extract just the key part from the record
+          const sponsor = sponsors[0].substring(sponsors[0].indexOf("=") + 1);
+
+          console.log(`${domainName} => ${skylink} | sponsor: ${sponsor}`);
+
```
Comment (interstella-5555) [inline src/resolver.js]:
```suggestion
```

File: src/resolver.js [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+          return resolve({ skylink, sponsor });
+        }
+
+        console.log(`${domainName} => ${skylink}`);
+
```
Comment (interstella-5555) [inline src/resolver.js]:
```suggestion
```

File: src/server.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+        res.json({ skylink, sponsor });
+      } catch (error) {
+        res.status(400).contentType("text/plain; charset=utf-8").send(error.message);
+      }
+    });
+  }
+
+  start(port, host) {
+    this.server = this.app.listen(port, (host = "0.0.0.0"), (error) => {
```
Comment (interstella-5555) [inline src/server.js]: I think it's more natural and a little bit less confusing to define a default argument value in the function declaration. In your code, `host` is undefined first and then the value changes when it's set inline so if we would add some code between, `host` would be undefined there.

```suggestion
  start(port, host = "0.0.0.0") {
    this.server = this.app.listen(port, host, (error) => {
```

**Review by interstella-5555 — CHANGES_REQUESTED** [review]
Let's also add 3 integration tests (domain with dnslink, domain with dnslink and sponsor, domain without dnslink) that make actual request to `http://<server>:<port>/dnslink`. We would expect the server to be running locally. We can set up either new domain for that or use one that we already own and set up subdomains.

File: src/server.js (hunk 8 above)
Comment (meeh0w) [reply src/server.js]: You're definitely right; that was my intention but I set it in a wrong place by accident (signatures are very similar).

As it is right now, passing `host` to `.start` call wouldn't even work, it would be overwritten.

File: src/resolver.js (hunk 6 above)
Comment (meeh0w) [reply src/resolver.js]: Oh, sure :)

File: package.json (hunk 1 above)
Comment (meeh0w) [reply package.json]: It gives us a nice intellisense though :)

File: package.json (hunk 2 above)
Comment (meeh0w) [reply package.json:14]: Version 3 only supports ESM format, I didn't want to switch away from CJS in this PR.

**Comment by meeh0w** [comment]
@MSevey what additional comments would you like to see? I feel like `describe` and `it` descriptions are pretty good there.

File: package.json (hunk 1 above)
Comment (interstella-5555) [reply package.json]: I think you can add it as a global package to your vscode if you want the intellisense (I've done that in the past but it might have been in webstorm and not vscode), otherwise I'd say it's unnecessarily included in the resulting image

Comment (meeh0w) [reply package.json]: On that note, we should probably run `yarn --prod` in our `Dockerfile`s so dev dependencies aren't be installed (I created [SKY-867](https://linear.app/skynetlabs/issue/SKY-867/run-yarn-in-production-mode)).

But fair enough, removed in [3f7dfd6](https://github.com/SkynetLabs/webportal-dnslink-api/pull/5/commits/3f7dfd6a54b14fc427e9cc53553c1bfd01d1b469)

File: .github/workflows/ci_release.yml [hunk 9]
```diff
@@ -36,6 +36,9 @@ jobs:
         uses: SkynetLabs/.github/.github/actions/yarn-basic-checks@master
         with:
           node-version: ${{ matrix.node-version }}
+      
+      - name: "Report test coverage impact"
```
Comment (MSevey) [inline .github/workflows/ci_release.yml:40]: nit: this could probably be a standalone job since it doesn't need the tests to run first right? or does it?

Comment (meeh0w) [reply .github/workflows/ci_release.yml:40]: It does, tests generate a report and place it in `coverage` directory.

**Comment by MSevey** [comment]
> @MSevey what additional comments would you like to see? I feel like `describe` and `it` descriptions are pretty good there.

Well I think it is a combination of a number of things.

First, the code itself has almost no comments. So it is very hard for reviews to review because you have to build all the context in your head from just reading the code. 

Since we have so many repos and so many of them have 1 maybe 2 key contributors, the reviewers often don't have the same level of understanding of the repository as the person writing the code. 

So even though the `describe` and `it` statements are relatively clear, it is hard for me to really understand if the testing is sufficient.

Part of this is also due to the fact that we are adding testing for an entire repo at one time that only you and karol have ever contributed to, so anyone else is at a pretty severe disadvantage in terms of reviewing the code. 

And again, this is all historical issues, not issues with this PR in particular. 

So in terms of comments that I would be looking for, it would first start with the code itself. The `Resolver` for example should have sufficient inline documentation that any reviewer and ramp up quickly on the intent of the code and understand the reasoning for why it is set up the way it is.

With that knowledge, reviewers can then better understand how the test coverage is working, and if there are any gaps that they see. 


100% code coverage from a code coverage report doesn't necessarily mean we have good code coverage. The a bad test that covers the code is not better than no tests at all.

File: src/logger.js [hunk 10]
```diff
@@ -0,0 +1,21 @@
+const winston = require("winston");
+
+const logger = winston.createLogger({
+  level: "info",
+  format: winston.format.json(),
+  defaultMeta: { service: "dnslink-api" },
+  transports: [
+    new winston.transports.File({ filename: "error.log", level: "error" }),
```
Comment (MSevey) [inline src/logger.js]: logfile names should be constants.

File: src/logger.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    new winston.transports.File({ filename: "log.log" }),
```
Comment (MSevey) [inline src/logger.js]: what is the reasoning behind having a separate error log and activity log?

File: src/resolver.js [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+const logger = require("./logger");
+
+const dnslinkNamespace = "skynet-ns";
+const sponsorNamespace = "skynet-sponsor-key";
+const dnslinkRegExp = new RegExp(`^dnslink=/${dnslinkNamespace}/.+$`);
+const sponsorRegExp = new RegExp(`^${sponsorNamespace}=[a-zA-Z0-9]+$`);
+const dnslinkSkylinkRegExp = new RegExp(`^dnslink=/${dnslinkNamespace}/([a-zA-Z0-9_-]{46}|[a-z0-9]{55})`);
+const hint = `valid example: dnslink=/${dnslinkNamespace}/3ACpC9Umme41zlWUgMQh1fw0sNwgWwyfDDhRQ9Sppz9hjQ`;
+
+class Resolver {
```
Comment (MSevey) [inline src/resolver.js:24]: All exported Classes and functions should have docstrings to explain there functionality

File: src/resolver.js [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+  async resolve(domainName) {
+    const lookup = `_dnslink.${domainName}`;
+
+    return new Promise((resolve) => {
+      dns.resolveTxt(lookup, (error, addresses) => {
```
Comment (MSevey) [inline src/resolver.js:42]: Ideally all these if conditions have comments explaining why we are checking for them.

The why and the context is important for others and reviewers to be able to onramp quickly.

File: src/resolver.js [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        logger.info(`${domainName} => ${skylink}`);
+
+        return resolve({ skylink });
+      });
+    });
+  }
+
+  validateRequest(req) {
```
Comment (MSevey) [inline src/resolver.js:138]: Example
```suggestion
  // validateRequest is a helper that wraps the isValidDomain check and returns an InvalidRequestError type.
  validateRequest(req) {
```

File: src/resolver.js [hunk 15]
```diff
@@ -0,0 +1,108 @@
...
+    if (!isValidDomain(req.params.name)) {
+      throw new InvalidRequestError(`${req.params.name} is not a valid domain`);
+    }
+  }
+}
+
+class InvalidRequestError extends Error {}
```
Comment (MSevey) [inline src/resolver.js]: Any syntax reason for defining these at the bottom of the file?

Typically we try and define all classes and types at the top of the file so that as you are reading the file you get all the class and type information first before reading the functions.

File: src/resolver.test.js [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+  const resolve = () => resolver.resolve("dummy-domain.com");
+
+  beforeEach(() => {
+    resolver = new Resolver();
+  });
+
+  describe(".validateRequest", () => {
+    it("throws on invalid domain", () => {
+      expect(() => resolver.validateRequest({ params: { name: "xyz" } })).toThrow(InvalidRequestError);
```
Comment (MSevey) [inline src/resolver.test.js]: Is there a way to run these tests against an array of test cases?  For valid and invalid test cases it is good to have a range of values to test against. 

Maybe something like

invalid
 - xyz
 - xyz.com.notvalid
 - .com
 - my.weird.domain.com.very.werid

valid
 - a.com
 - xyz.com
 - sub.domain.com
 - xyz.net

File: src/server.js [hunk 17]
```diff
@@ -0,0 +1,40 @@
+const express = require("express");
+
+const logger = require("./logger");
+const { Resolver } = require("./resolver");
+
+class Server {
```
Comment (MSevey) [inline src/server.js:9]: Exported Classes should have docstrings

File: src/server.js [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+  configure() {
+    this.app.get("/dnslink/:name", async (req, res) => {
+      const domainName = req.params.name;
+
+      try {
+        this.resolver.validateRequest(req);
+
+        const { skylink, sponsor } = await this.resolver.resolve(domainName);
+
+        res.json({ skylink, sponsor });
```
Comment (MSevey) [inline src/server.js:28]: Here is a good example of why documentation is important. Even after reading through the code it is not clear to me what the `sponsor` value is since the `Resolver` doesn't have documentation about what the return arguments are.

File: src/server.test.js [hunk 19]
```diff
@@ -13,6 +13,40 @@ describe("Server", () => {
     app.server.close();
   });
 
+  describe("- live requests -", () => {
```
Comment (interstella-5555) [inline src/server.test.js]: I would suggest moving integration tests into a separate file for visibility, the idea is that they do not test a particular file in the code but the whole project end to end.

Let's also add one test that checks for specific error handling like malformed skylink.

File: src/logger.js (hunk 10 above)
Comment (interstella-5555) [inline src/logger.js]: We should put logs in a `logs` directory to be able to easily mount that directory in docker-compose.

File: src/logger.js (hunk 11 above)
Comment (interstella-5555) [inline src/logger.js]: I think we could use Console transport the non-error logs, we don't really need those persisted. For errors I would like them to be persisted like you set them for possible debugging.

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: src/logger.js (hunk 11 above)
Comment (meeh0w) [reply src/logger.js]: Just a convention I got used to. Karol mentioned, we can get by with persisting only error logs, so I removed the activity log file.

File: src/server.js (hunk 18 above)
Comment (meeh0w) [reply src/server.js:28]: I added JSdocs in [406df08](https://github.com/SkynetLabs/webportal-dnslink-api/pull/5/commits/406df08abb0e4fd10329083d1722779c6a7623eb) :)

File: src/server.e2e.test.js [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+    const expectedResponse = {
+      skylink: "MABdWWku6YETM2zooGCjQi26Rs4a6Hb74q26i-vMMcximQ",
+      sponsor: "JDCLOJIJ7NJRSN4QRD7EDFVNP7MPJ24O856D57NE3FV2PFBAT6T0",
+    };
+
+    expect(response.status).toBe(200);
+    expect(await response.json()).toEqual(expectedResponse);
+  });
+
+  it("dns-link.skynetlabs.io", async () => {
```
Comment (interstella-5555) [inline src/server.e2e.test.js]:
```suggestion
  it("dns-link.skynetlabs.io - dnslink entry correct with valid skylink", async () => {
```

File: src/server.e2e.test.js [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+    const expectedResponse = {
+      skylink: "MABdWWku6YETM2zooGCjQi26Rs4a6Hb74q26i-vMMcximQ",
+      sponsor: "JDCLOJIJ7NJRSN4QRD7EDFVNP7MPJ24O856D57NE3FV2PFBAT6T0",
+    };
+
+    expect(response.status).toBe(200);
+    expect(await response.json()).toEqual(expectedResponse);
+  });
+
+  it("malformed-dns-link.skynetlabs.io", async () => {
```
Comment (interstella-5555) [inline src/server.e2e.test.js]:
```suggestion
  it("malformed-dns-link.skynetlabs.io - dns entry correct but skylink invalid", async () => {
```

File: src/server.e2e.test.js [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+    const response = await fetch("http://0.0.0.0:1235/dnslink/dns-link.skynetlabs.io");
+    const expectedResponse = {
+      skylink: "MABdWWku6YETM2zooGCjQi26Rs4a6Hb74q26i-vMMcximQ",
+    };
+
+    expect(response.status).toBe(200);
+    expect(await response.json()).toEqual(expectedResponse);
+  });
+
+  it("sponsored-dns-link.skynetlabs.io", async () => {
```
Comment (interstella-5555) [inline src/server.e2e.test.js]:
```suggestion
  it("sponsored-dns-link.skynetlabs.io - dnslink entry correct with valid skylink and sponsor key", async () => {
```

File: src/server.e2e.test.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+  it("no-dns-link.skynetlabs.io", async () => {
+    const response = await fetch("http://0.0.0.0:1235/dnslink/sponsored-dns-link.skynetlabs.io");
+    const expectedResponse = {
+      skylink: "MABdWWku6YETM2zooGCjQi26Rs4a6Hb74q26i-vMMcximQ",
+      sponsor: "JDCLOJIJ7NJRSN4QRD7EDFVNP7MPJ24O856D57NE3FV2PFBAT6T0",
+    };
+
+    expect(response.status).toBe(200);
+    expect(await response.json()).toEqual(expectedResponse);
+  });
```
Comment (interstella-5555) [inline src/server.e2e.test.js]: I assume that should be a test that checks whether a domain without dnslink configured gets properly rejected.

```suggestion
  it("no-dns-link.skynetlabs.io - no dnslink entry", async () => {
    const response = await fetch("http://0.0.0.0:1235/dnslink/no-dns-link.skynetlabs.io");

    expect(response.status).toBe(400);
    expect(response.headers.get("content-type")).toEqual("text/plain; charset=utf-8");
    expect(await response.text()).toEqual("ENOTFOUND: _dnslink.no-dns-link.skynetlabs.io TXT record doesn't exist");
  });
```

**Review by interstella-5555 — CHANGES_REQUESTED** [review]

File: src/resolver.js (hunk 15 above)
Comment (meeh0w) [reply src/resolver.js]: Not really, I tend to do that when the class in question doesn't really add any specific logic (mostly errors).

**Review by interstella-5555 — DISMISSED** [review]

**Review by interstella-5555 — APPROVED** [review]

**Review by MSevey — APPROVED** [review]
